File: Year 2/Semester 4/MPP/JavaFourthWeek/src/main/java/repository/ArtistDBRepository.java
package repository;

import domain.Artist;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ArtistDBRepository extends AbstractRepository<Integer, Artist> {

    public ArtistDBRepository(Properties props) {
        super(props);
    }

    private Artist getArtistFromResultSet(ResultSet resultSet) throws SQLException {
        int id = resultSet.getInt("id");
        String firstName = resultSet.getString("first_name");
        String lastName = resultSet.getString("last_name");
        return new Artist(id, firstName, lastName);
    }

    @Override
    public Artist findById(Integer id) {
        logger.traceEntry("find an artist by id");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "select * from artist where id = ?")) {
            ps.setInt(1, id);
            try (ResultSet resultSet = ps.executeQuery()) {
                if (resultSet.next()) {
                    Artist artist = getArtistFromResultSet(resultSet);
                    logger.traceExit(artist);
                    return artist;
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("ERROR for findById in ArtistDBRepository: " + exception);
        }
        return null;
    }

    @Override
    public Iterable<Artist> getAll() {
        logger.traceEntry("finding all artists");
        Connection connection = dbUtils.getConnection();
        List<Artist> artists = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement("select * from artist")) {
            try (ResultSet resultSet = ps.executeQuery()) {
                while (resultSet.next()) {
                    artists.add(getArtistFromResultSet(resultSet));
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("ERROR for getAll in ArtistDBRepository: " + exception);
        }
        logger.traceExit(artists);
        return artists;
    }

    @Override
    public boolean add(Artist artist) {
        logger.traceEntry("adding artist {}", artist);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "insert into artist (first_name, last_name) values (?, ?)")) {
            ps.setString(1, artist.getFirstName());
            ps.setString(2, artist.getLastName());
            int result = ps.executeUpdate();
            logger.trace("saved {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for insert in ArtistDBRepository: " + exception);
            return false;
        }
        logger.traceExit("inserted successfully");
        return true;
    }

    @Override
    public boolean delete(Integer id) {
        logger.traceEntry("delete artist");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("delete from artist where id=?")) {
            ps.setInt(1, id);
            int result = ps.executeUpdate();
            logger.trace("deleted {} instance", result);
            if (result == 0)
                return false;
        } catch (SQLException exception) {
            logger.error("ERROR for delete in ArtistDBRepository: " + exception);
            return false;
        }
        logger.traceExit("deleted successfully");
        return true;
    }

    @Override
    public boolean update(Artist artist) {
        logger.traceEntry("updating artist");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "update artist set first_name = ?, last_name = ?" +
                " where id = ?")) {
            ps.setString(1, artist.getFirstName());
            ps.setString(2, artist.getLastName());
            ps.setInt(3, artist.getId());
            int result = ps.executeUpdate();
            logger.trace("updated {} instance", result);
            if (result == 0)
                return false;
        } catch (SQLException exception) {
            logger.error("ERROR for update in ArtistDBRepository: " + exception);
            return false;
        }
        logger.traceExit("updated successfully");
        return true;
    }
}

File: Year 2/Semester 4/MPP/ProtobuffProjectCSharp/persistence/ArtistDBRepository.cs
using log4net;
using ProtobuffProject.domain;
using ProtobuffProject.repository.interfaces;
using System.Collections.Generic;
using System.Data.SQLite;

namespace ProtobuffProject.repository
{
    public class ArtistDBRepository: IArtistRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("ArtistDBRepository");

        private readonly IDictionary<string, string> props;

        public ArtistDBRepository(IDictionary<string, string> props)
        {
            this.props = props;
            Logger.Info("Creating ArtistDBRepository");
        }

        private static Artist<int> GetArtistFromResultSet(SQLiteDataReader resultSet)
        {
            var id = resultSet.GetInt32(0);
            var firstName = resultSet.GetString(1);
            var lastName = resultSet.GetString(2);
            return new Artist<int>(id, firstName, lastName);
        }
        
        public List<Artist<int>> GetAll() {
            Logger.Info("finding all artists");
            var connection = DBUtils.GetConnection(props);
            var list = new List<Artist<int>>();
            using (var cmd = new SQLiteCommand("select * from artist", connection as SQLiteConnection))     
                using (var reader = cmd.ExecuteReader())
                    while (reader.Read())
                        list.Add(GetArtistFromResultSet(reader));
            Logger.Info(list);
            return list;
        }
            

        public Artist<int> FindById(int id)
        {
            Logger.Info("find an artist by id");
            var connection = DBUtils.GetConnection(props);
            var cmd = new SQLiteCommand("select * from artist where id = " + id,
                connection as SQLiteConnection);
            using (var reader = cmd.ExecuteReader())
            {
                if (!reader.Read()) return null;
                var artist = GetArtistFromResultSet(reader);
                Logger.Info(artist);
                return artist;
            }
        }

        public bool Add(Artist<int> artist)
        {
            Logger.Info("adding artist");
            var connection = DBUtils.GetConnection(props);
            using (var command = new SQLiteCommand
                ("insert into artist values (@id, @first_name, @last_name)",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", null);
                command.Parameters.AddWithValue("@first_name", artist.FirstName);
                command.Parameters.AddWithValue("@last_name", artist.LastName);
                command.ExecuteNonQuery();
                Logger.InfoFormat("inserted successfully {0}", artist);
                return true;
            }
        }


        public bool Delete(int id)
        {
            Logger.InfoFormat("delete artist with id = {0}", id);
            var connection = DBUtils.GetConnection(props);
            using (var command = new SQLiteCommand("delete from artist where id = @id",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", id);
                command.ExecuteNonQuery();
                Logger.InfoFormat("deleted successfully for id {0}", id);
                return true;
            }
        }

        public bool Update(Artist<int> artist)
        {
            Logger.Info("updating artist");
            var connection = DBUtils.GetConnection(props);
            using (var command = new SQLiteCommand("update artist set first_name = @first_name, " +
                "last_name = @last_name where id = @id", connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@first_name", artist.FirstName);
                command.Parameters.AddWithValue("@last_name", artist.LastName);
                command.Parameters.AddWithValue("@id", artist.Id);
                command.ExecuteNonQuery();
                Logger.Info("updated successfully");
                return true;
            }
        }
    }
}
